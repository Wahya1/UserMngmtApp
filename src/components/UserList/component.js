import React from "react";
import { Container, Grid, Typography } from "@mui/material";
import { useSelector } from "react-redux";
import UserItems from "../UserItems/component";

const UserList = () => {
  const { loading, users, error } = useSelector((state) => state.user);

  if (loading) return <Typography variant="h6">Chargement...</Typography>;
  if (error)
    return (
      <Typography variant="h6" color="error">
        Erreur : {error}
      </Typography>
    );

  return (
    <Container sx={{ mb: 10 }}>
      <Typography
        variant="h6"
        sx={{
          mb: 4,
          fontSize: "1.8rem",
          fontWeight: 600,
          color: "black",
          letterSpacing: "0.5px",
          textAlign: "center",
          textTransform: "uppercase",
        }}
      >
        Users List
      </Typography>
      <Grid container spacing={2}>
        {users.map((user) => (
          <UserItems
            key={user.id}
            id={user.id}
            name={user.name}
            username={user.username}
            email={user.email}
            address={user.address}
            phone={user.phone}
          />
        ))}
      </Grid>
    </Container>
  );
};

export default UserList;
