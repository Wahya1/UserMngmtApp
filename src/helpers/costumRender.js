import { Provider } from "react-redux";
import { render } from "@testing-library/react";

import store from "../test-utils/setupTestStore";

export const customRender = (comp) =>
  render(<Provider store={store}>{comp}</Provider>);
