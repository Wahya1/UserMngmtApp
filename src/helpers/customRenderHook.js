import { renderHook } from "@testing-library/react";
import store from "../test-utils/setupTestStore";
import { Provider } from "react-redux";

export const customRenderHook = (hook) =>
  renderHook(hook, {
    wrapper: ({ children }) => <Provider store={store}>{children}</Provider>,
  });
