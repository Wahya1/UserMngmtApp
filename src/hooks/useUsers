import { useDispatch, useSelector } from "react-redux";
import useDispatchedAction from "./useDispatchedAction";
import { fetchUsersFailure } from "../redux/UserReducer/action";

const useUsers = () => {
  const usersData = useSelector((store) => store.user);

  const handleFetchUsersRequest = () => useDispatchedAction(fetchUsersRequest);
  const handleFetchUsersSuccess = () => useDispatchedAction(fetchUsersSuccess);
  const handleFetchUsersFailure = () => useDispatchedAction(fetchUsersFailure);

  useEffect(() => {
    handleFetchUsersRequest();

    fetch("https://jsonplaceholder.typicode.com/users")
      .then((response) => {
        if (response.ok) return response.json();
        throw new Error("error when users fetching ");
      })
      .then((data) => {
        const mappedData = data.slice(0, 6);
        handleFetchUsersSuccess(mappedData);
      })
      .catch((error) => {
        handleFetchUsersFailure(error.message);
      });
  }, [dispatch]);

  return { usersData };
};

export default useUsers;
