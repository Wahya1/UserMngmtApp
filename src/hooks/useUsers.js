import { useSelector } from "react-redux";
import useDispatchedAction from "./useDispatchedAction";

import {
  fetchUsersFailure,
  fetchUsersRequest,
  fetchUsersSuccess,
} from "../redux/UsersReducer/action";

import { useEffect } from "react";

const useUsers = () => {
  const usersData = useSelector((store) => store.users.users);
  console.log("usersdata:", { usersData });

  const handleFetchUsersRequest = useDispatchedAction(fetchUsersRequest);
  const handleFetchUsersSuccess = useDispatchedAction(fetchUsersSuccess);
  const handleFetchUsersFailure = useDispatchedAction(fetchUsersFailure);

  useEffect(() => {
    handleFetchUsersRequest();

    fetch("https://jsonplaceholder.typicode.com/users")
      .then((response) => {
        if (response.ok) return response.json();
        throw new Error("error when users fetching ");
      })
      .then((data) => {
        console.log("mapped data ::::::::::::::::::::::");
        const mappedData = data.slice(0, 6);
        handleFetchUsersSuccess(mappedData);
      })
      .catch((error) => {
        handleFetchUsersFailure(error.message);
      });
  }, []);

  return usersData;
};

export default useUsers;
