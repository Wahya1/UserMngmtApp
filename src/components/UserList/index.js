import { useDispatch } from "react-redux";
import React, { useEffect } from "react";

import { Box } from "@mui/material";

import UserList from "./component";
import AddUserDialogContainer from "../AddUserDialog";

import {
  fetchUsersFailure,
  fetchUsersRequest,
  fetchUsersSuccess,
} from "../../redux/UserReducer/action";

const UserListContainer = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchUsersRequest());

    fetch("https://jsonplaceholder.typicode.com/users")
      .then((response) => {
        if (response.ok) return response.json();
        throw new Error("error when users fetching ");
      })
      .then((data) => {
        dispatch(fetchUsersSuccess(data));
      })
      .catch((error) => {
        dispatch(fetchUsersFailure(error.message));
      });
  }, [dispatch]);

  return (
    <Box sx={{ padding: 2 }}>
      <AddUserDialogContainer />
      <UserList />
    </Box>
  );
};

export default UserListContainer;
