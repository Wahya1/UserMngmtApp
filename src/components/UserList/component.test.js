import { useSelector } from "react-redux";
import useUsers from "../../hooks/useUsers";
import UserList from "./component";
import { screen } from "@testing-library/react";
import { customRender } from "../../helpers/costumRender";

jest.mock("../../hooks/useUsers");

jest.mock("react-router", () => ({
  ...jest.requireActual("react-router"),
  useNavigate: jest.fn(),
  useParams: jest.fn(() => ({})),
}));

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
}));

describe("UserList component", () => {
  it("displays loading state", () => {
    useUsers.mockReturnValue([]);
    useSelector.mockReturnValue({ loading: true, error: null });

    const container = customRender(<UserList />);
    expect(container.asFragment()).toMatchSnapshot();
    expect(screen.getByText("Chargement...")).toBeInTheDocument();
  });

  it("displays erreur when error ", () => {
    useUsers.mockReturnValue([]);
    useSelector.mockReturnValue({
      loading: false,
      error: "lors du chargement des donnees",
    });

    customRender(<UserList />);
    expect(
      screen.getByText("Erreur : lors du chargement des donnees")
    ).toBeInTheDocument();
  });

  test("displays user list", () => {
    useUsers.mockReturnValue([
      {
        id: 1,
        name: "Driss",
        username: "driss123",
        email: "driss@example.com",
        address: { city: "Paris" },
        phone: "1234567890",
      },
    ]);

    useSelector.mockReturnValue({ loading: false, error: null });

    customRender(<UserList />);

    expect(screen.getByText("Users List")).toBeInTheDocument();
  });
});

// import React from "react";
// import { render, screen, fireEvent } from "@testing-library/react";
// import UserItems from "./component";
// import { useDispatch } from "react-redux";
// import { useNavigate } from "react-router";
// import { deleteUser } from "../../redux/UsersReducer/action";

// jest.mock("react-redux", () => ({
//   useDispatch: jest.fn(),
// }));

// jest.mock("../../redux/UsersReducer/action", () => ({
//   deleteUser: jest.fn(),
// }));

// describe("UserItems Component", () => {
//   const mockDispatch = jest.fn();
//   const mockNavigate = jest.fn();

//   beforeEach(() => {
//     useDispatch.mockReturnValue(mockDispatch);
//     useNavigate.mockReturnValue(mockNavigate);
//   });

//   it("calls handleDelete when delete button is clicked", () => {
//     const userProps = {
//       id: 1,
//       name: "John Doe",
//       username: "johnd",
//       email: "john@example.com",
//       address: { city: "New York" },
//       phone: "123-456-7890",
//     };

//     const container = render(<UserItems {...userProps} />);
//     console.log(container.log());

//     // const deleteButton = screen.getByLabelText(/delete/i);
//     // fireEvent.click(deleteButton);

//     expect(mockDispatch).toHaveBeenCalledWith(deleteUser(1)); // Check if deleteUser is called with correct id
//   });
// });
