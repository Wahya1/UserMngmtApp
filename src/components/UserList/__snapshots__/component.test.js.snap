// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList component displays loading state 1`] = `
<DocumentFragment>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
  >
    Chargement...
  </h6>
</DocumentFragment>
`;
